connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -0,0 +1,585 @@
+    }
+
+    /**
+     * If version is V2, then paths are cached for further access.
+     *
+     * @param field   field path expression
+     * @param version field syntax version
+     */
+    public static SingleFieldPath of(String field, FieldSyntaxVersion version) {
+        if ((field == null || field.isEmpty()) // empty path
+                || version.equals(FieldSyntaxVersion.V1)) { // or V1
+            return new SingleFieldPath(field, version);
+        } else { // use cache when V2
+            final SingleFieldPath found = PATHS_CACHE.get(field);
+            if (found != null) {
+                return found;
+            } else {
+                final SingleFieldPath fieldPath = new SingleFieldPath(field, version);
+                PATHS_CACHE.put(field, fieldPath);
+                return fieldPath;
+            }
+        }
+    }
+
+    SingleFieldPath(String pathText, FieldSyntaxVersion version) {
+        if (pathText == null || pathText.isEmpty()) { // empty path
+            this.path = new String[] {};
+        } else {
+            switch (version) {
+                case V1: // backward compatibility
+                    this.path = new String[] {pathText};
+                    break;
+                case V2:
+                    path = buildFieldPathV2(pathText);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown syntax version: " + version);
+            }
+        }
+    }
+
+    private String[] buildFieldPathV2(String pathText) {
+        // if no dots or wrapping backticks are used, then return path with single step
+        if (!pathText.contains(String.valueOf(DOT_CHAR))) {
+            return new String[] {pathText};
+        } else {
+            // prepare for tracking path steps
+            final List<String> steps = new ArrayList<>();
+            // avoid creating new string on changes
+            final StringBuilder s = new StringBuilder(pathText);
+
+            while (s.length() > 0) { // until path is traversed
+                // start processing backtick pair, if any
+                if (s.charAt(0) == BACKTICK_CHAR) {
+                    s.deleteCharAt(0);
+
+                    // find backtick closing pair
+                    int idx = 0;
+                    while (idx >= 0) {
+                        idx = s.indexOf(String.valueOf(BACKTICK_CHAR), idx);
+                        if (idx == -1) { // if not found, fail
+                            throw new IllegalArgumentException("Incomplete backtick pair at [...]`" + s);
+                        }
+                        // check that it is not escaped or wrapped in another backticks pair
+                        if (idx < s.length() - 1 // not wrapping the whole field path
+                                && (s.charAt(idx + 1) != DOT_CHAR // not wrapping
+                                || s.charAt(idx - 1) == BACKSLASH_CHAR)) { // ... or escaped
```

Comment:
We can make this a little bit easier to read by pulling out all of the sub-conditions of the `if` statement into their own variables:
```suggestion
                    boolean endOfPath = idx >= s.length() - 1;
                    boolean notWrappingField = !endOfPath && s.charAt(idx + 1) != DOT;
                    boolean escaped = s.charAt(idx - 1) == BACKSLASH;
                    // check that it is not escaped or wrapped in another backticks pair
                    if (!endOfPath && (notWrappingField || escaped)) {
```